src/method/method_invoke_group.cpp

```diff
@@ -219,28 +237,70 @@ namespace cubmethod
 
     pr_clear_value_vector (m_result_vector);
 
-    for (method_invoke *method: m_method_vector)
-      {
-	method->reset (m_thread_p);
-      }
+    // destroy cursors used in this group
+    destory_all_cursors ();
 
-    if (!is_end_query)
-      {
-	cubmethod::header header (METHOD_REQUEST_END, get_id());
-	error = method_send_data_to_client (m_thread_p, header);
-      }
+    cubmethod::header header (METHOD_REQUEST_END, get_id());
+    error = method_send_data_to_client (m_thread_p, header);
 
     return error;
   }
 
-  int method_invoke_group::end ()
+  void
+  method_invoke_group::end ()
   {
-    int error = NO_ERROR;
+    if (m_is_running == false)
+      {
+	return;
+      }
+
     reset (true);
 
     get_connection_pool ()->retire (m_connection);
     m_connection = nullptr;
 
-    return error;
```

Comment:
not necessary to check m_connection is null at "retire"?

Reply:
retire returns immediately if m_connection is null and does nothing